docs/decex/fraud_proofs.md

```diff
@@ -8,21 +8,32 @@ keywords:
+Every domain operator executes the domain block ([as described](workflow.md#domain-block-execution-on-the-operator-node)), derived deterministically from the consensus block, and submits this computational result to the consensus chain as an execution receipt within the next bundle this operator produces, thereby committing to the execution result. By default, the computation result is optimistically assumed correct until challenged by a fraud proof during the challenge period of `BlockTreePruningDepth` blocks. All domain nodes scrutinize the submitted execution results, and upon detecting any discrepancies, they challenge the execution by submitting a fraud proof to the consensus chain as an unsigned extrinsic. 
+A fraud proof either explicitly includes all necessary data and the state of the domain required for the verification process or points to a storage key for this data. This way, a fraud can be executed by a node on the consensus chain, which has access to the recent consensus state and the MMR for historical state, but not the domain state. If the node who detected fraud is also a registered operator of this domain, they can submit a new execution receipt to the consensus chain with their next bundle, which will override the fraudulent one in the [block tree](interfaces.md#block_tree) once the fraud proof is accepted by the consensus chain.
 
 Any domain node (a node that has an up-to-date state of domain A) can submit fraud proofs for domain A. Whether the node is acting honestly or not in this particular instance is determined by the validity of the fraud proof. The node does not have to stake or run operator (produce bundles) to report fraud.
 
 Fraud proofs are verified on the client side, first in the transaction pool and then on importing the block.
 
 Broadly, fraud proofs can be categorized into those caused by invalid execution receipt fields and those caused by invalid state transitions.
 
-We currently handle the following types of fraud proofs:
+## Structure
+
+A fraud proof contains the following fields:
+
+- `domain_id`: domain identifier of the domain this fraud proof targets
+- `bad_receipt_hash`: hash of the claimed invalid execution receipt
+- `maybe_mmr_proof`: MMR proof for the consensus state root that used to verify the storage proof. Only the  Invalid Extrinsics Root, Incorrect List of Inboxed Bundles fraud proofs are using the MMR proof during verification, for other fraud proofs it is set to `None`. 
+- `maybe_domain_runtime_code_proof`: domain runtime code storage proof and MMR proof for the parent block. May be set to `None` if the specific fraud proof variant doesn't require domain runtime code for verification or the required domain runtime code is available from the current runtime state. This field is usually required if the domain runtime code was upgraded during the challenge period.
```

Comment:
This seems problematic. Domain runtime code is unbounded. If such storage is included in the Fraud proof, then FP  size is unbounded as well.
This could be an issue if the FP doesn't fit in to the block due to size restrictions.

Instead, FP should ideally hold the hash of the domain runtime that is required and we can hold all the runtime upgrades of the domain in the consensus state itself. We do not necessarily hold all of them but rather hold those runtimes for which domain blocks are still in the Challenge period.

Reply:
> Domain runtime code is unbounded

This seems deeply problematic as well

Reply:
> This may not be true, as the domain runtime code register/upgrade required a sudo consensus extrinsic (except the genesis domain runtime), if the runtime code is unbounded then we can't register/upgrade the runtime in the first place.

Well, while registering a domain runtime there is no extra data that is passed over unlike the fraud proof where we have other storage proofs. While we can limit sizes of each storage proofs, there is no necessary gurantee that all combined will fit into the block unfortunately.

> Also, notice that fraud proof is expected to be rare, even under attack it happens more frequently it is not sustainable.

Irrespective whether a Fraud proof is rare or not, we should be able to gurantee the inclusion of such fraud proof with domain runtime. While the current runtimes may or may not fit the block, this approach may solve the problem until we have bigger runtimes. They we will see the same issue of not inclusion.

This is problem is even more pronounced when the we allow open domain runtimes and might not be so useful to have such a proof in FP.  

While the size the storage proof from reads may seem unbounded, this is could solved with some changes albiet maybe invasive but we cannot reduce the size of runtime itself.

Reply:
> While we can limit sizes of each storage proofs, there is no necessary gurantee that all combined will fit into the block unfortunately.

The domain runtime code storage proof is the largest, other storage proofs are only a few KB (and are mostly constant), my only concern is the execution proof used in the invalid state transition FP and the illegal tx FP but it is guarded by the excessive transaction limit (hope it works).

A domain runtime code is 400-600 KB while a consensus block is 5MB if still a fraud proof can't fit into the block then the real problem we should think about is where the >4MB of space is used.

>While the current runtimes may or may not fit the block, this approach may solve the problem until we have bigger runtimes. They we will see the same issue of not inclusion.
>This is problem is even more pronounced when the we allow open domain runtimes and might not be so useful to have such a proof in FP.

FYI, the [polkadot runtime](https://github.com/paritytech/polkadot-sdk/releases/tag/polkadot-v1.7.0) with >60 pallets is 1.6MB (after compressed).

If we store all the unupgraded domain runtime in the state then another problem is with open domain/runtime what if some domains just not upgrade intentionally or worse keep uploading new versions of runtime to the consensus chain to bloat the state?

Overall, I don't think the domain runtime code storage proof is an issue for the future that I can foresee.